src/utils/XFDR.cpp

```diff
@@ -0,0 +1,1095 @@
+    }
+
+    /**
+     * @brief Prepares read vector of PeptideIdentification such that it can be further processed downstream
+     * @param pep_ids vector of PeptideIdentification to be prepared
+     */
+    void prepareInput(vector< PeptideIdentification > & pep_ids)
+    {
+      for (vector< PeptideIdentification >::iterator pep_ids_it = pep_ids.begin();
+           pep_ids_it != pep_ids.end(); ++pep_ids_it)
+      {
+        PeptideIdentification & pep_id = *pep_ids_it;
+
+        double score = getXLMetaValue<double>("OpenXQuest:score", pep_id, true);
+
+        if (score < this->min_score)
+        {
+          this->min_score = std::floor(score);
+        }
+        if (score > this->max_score)
+        {
+          this->max_score = std::ceil(score);
+        }
+
+        const vector< PeptideHit > & pep_hits = pep_id.getHits();
+        // Set peptide identification to target or decoy
+        if (    pep_hits[0].getMetaValue("target_decoy").toString() == "decoy"
+                || (pep_hits.size() == 2 && pep_hits[1].getMetaValue("target_decoy").toString() == "decoy"))
+        {
+          pep_id.setMetaValue("target_decoy", DataValue("decoy"));
+        }
+        else
+        {
+          pep_id.setMetaValue("target_decoy", DataValue("target"));
+        }
+
+        // figure out if cross-link is inter- or intra protein
+        if (pep_hits.size() == 2)
+        {
+          vector< PeptideEvidence > alpha_ev = pep_hits[0].getPeptideEvidences();
+          vector< PeptideEvidence > beta_ev = pep_hits[1].getPeptideEvidences();
+
+          for (vector< PeptideEvidence >::const_iterator alpha_ev_it  = alpha_ev.begin();
+               alpha_ev_it != alpha_ev.end(); ++alpha_ev_it)
+          {
+            for (vector< PeptideEvidence >::const_iterator beta_ev_it = beta_ev.begin();
+                 beta_ev_it != beta_ev.end(); ++beta_ev_it)
+            {
+              String alpha_prot = alpha_ev_it->getProteinAccession();
+              String beta_prot = beta_ev_it->getProteinAccession();
+
+              Internal::XQuestResultXMLHandler::removeSubstring(alpha_prot, "reverse_");
+              Internal::XQuestResultXMLHandler::removeSubstring(alpha_prot, Internal::XQuestResultXMLHandler::decoy_string);
+              Internal::XQuestResultXMLHandler::removeSubstring(beta_prot, "reverse_");
+              Internal::XQuestResultXMLHandler::removeSubstring(beta_prot, Internal::XQuestResultXMLHandler::decoy_string);
+              pep_id.setMetaValue( alpha_prot == beta_prot ? "OpenXQuest:is_intraprotein" : "OpenXQuest:is_interprotein" , DataValue());
+            }
+          }
+        }
+      }
+    }
+
+    /**
+     * @brief Add class to the scores Map, of not already present
+     * @param scores Map where the empty xl class should be added to
+     * @param name name of the xl class to be added
+     */
+    inline static void addEmptyClass(std::map< String, vector< double> > & scores, const String & name)
```

Comment:
it should throw an error if name already present! Otherwise, you may end up with very nasty stuff to debug